internal/broker/broker.go

```diff
@@ -636,6 +631,44 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 	return AuthGranted, userInfoMessage{UserInfo: authInfo.UserInfo}
 }
 
+// userNameIsAllowed checks whether the user's username is allowed to access the machine.
+func (b *Broker) userNameIsAllowed(userName string) bool {
+	// The user is allowed to login if:
+	// - ALL users are allowed
+	// - the user's name is in the list of allowed_users
+	// - the user is the owner of the machine and OWNER is in the allowed_users list
+	if b.cfg.userConfig.allUsersAllowed {
+		return true
+	} else if _, ok := b.cfg.userConfig.allowedUsers[userName]; ok {
+		return true
+	} else if b.cfg.userConfig.ownerAllowed {
+		// If owner is undefined, then the first user to login is considered the owner
+		if b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName {
+			return true
+		}
+	}
+	return false
+}
+
+// allowUser checks whether the user's username is allowed and registers the owner if necessary.
+func (b *Broker) allowUser(userName string) bool {
+	allowed := b.userNameIsAllowed(userName)
+
+	// If the owner is unset and allowed, we auto-generate a config file with the first
+	// user to login as the owner.
+	if allowed && b.cfg.userConfig.ShouldRegisterOwner() {
+		err := b.cfg.RegisterOwner(b.cfg.ConfigFile, userName)
```

Comment:
I wouldn't expect a method called `allowUser` to also register the user as the owner. I would prefer to move this to the caller (and then we can remove the `allowUser` method and let the caller call `userNameIsAllowed`  directly):

```diff
diff --git a/internal/broker/broker.go b/internal/broker/broker.go
index d33d794..c86c6bf 100644
--- a/internal/broker/broker.go
+++ b/internal/broker/broker.go
@@ -611,10 +611,22 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 		}
 	}
 
-	if !b.allowUser(authInfo.UserInfo.Name) {
+	if !b.userNameIsAllowed(authInfo.UserInfo.Name) {
 		return AuthDenied, errorMessage{Message: "permission denied"}
 	}
 
+	// If the owner is unset and allowed, we auto-generate a config file with the first
+	// user to login as the owner.
+	if b.cfg.userConfig.ShouldRegisterOwner() {
+		err := b.cfg.RegisterOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
+		if err != nil {
+			// The user is not allowed, if we fail to create the owner-autoregistration file.
+			// Otherwise the owner might change if the broker is restarted.
+			slog.Error(fmt.Sprintf("Failed to assign the owner role: %v", err))
+			return AuthDenied, errorMessage{Message: "could not assign the owner role"}
+		}
+	}
+
 	if session.isOffline {
 		return AuthGranted, userInfoMessage{UserInfo: authInfo.UserInfo}
 	}
@@ -650,25 +662,6 @@ func (b *Broker) userNameIsAllowed(userName string) bool {
 	return false
 }
 
-// allowUser checks whether the user's username is allowed and registers the owner if necessary.
-func (b *Broker) allowUser(userName string) bool {
-	allowed := b.userNameIsAllowed(userName)
-
-	// If the owner is unset and allowed, we auto-generate a config file with the first
-	// user to login as the owner.
-	if allowed && b.cfg.userConfig.ShouldRegisterOwner() {
-		err := b.cfg.RegisterOwner(b.cfg.ConfigFile, userName)
-		if err != nil {
-			// The user is not allowed, if we fail to create the owner-autoregistration file.
-			// Otherwise the owner might change if the broker is restarted.
-			slog.Error(fmt.Sprintf("Failed to auto-register the owner: %f", err))
-			return false
-		}
-	}
-
-	return allowed
-}
-
 func (b *Broker) startAuthenticate(sessionID string) (context.Context, error) {
 	session, err := b.getSession(sessionID)
 	if err != nil {
```

Reply:
The reason for not placing this logic in `handleIsAuthenticated` is that I didn't want to make an already huge function (that IMO is too complicated), even more more complicated. But since that is what you want, I will move it.

Reply:
I agree that the function is too big, in my opinion it should be refactored (and I plan to do that at some point). However, I think the

```
if b.cfg.userConfig.ShouldRegisterOwner() {
    err := b.cfg.RegisterOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
   // error handling
```

already adds as little complexity as possible without losing clarity.